[Feature] Add RoF2 Guild features

neckkola opened the pull request:
Good afternoon.

Over the past several months I have been working on adding the RoF2 Guild Features.  While it still requires some spit and polish, I believe it is ready for review.  I need to clean it up, rewrite some functions, so please be gentle.  :-)

I did not re-write everything as I tried to use the existing guild code as-is.  That said, happy to update as you see fit, particularly using the new patterns and repositories.  I learned a lot from my last PR that needs to be implemented here.  I don't want to wait any longer though, so thought I would push it today.
 
This adds the following guild features and design pattern the existing guild system;

- guild features are based on RoF2 within source with translators used to converted between client differences
- backward compatible with Ti and UF, and allows for mixed client servers
- Guild Bank for Ti and UF is based on RoF2 Permissions for banking so that the Guild Leader does not have to use RoF2 Client
- Guild Ranks and Permissions are enabled.
- Guild Tributes are enabled.
- Event logging via rules for donating tribute items and plat
- Rules to limit Guild Tributes based on max level of server
- Rewrote guild communications to client using specific opcodes
-- Server no longer sends a guild member list on each zone
-- Guild window is updated when a member levels, rank changes, zone changes, banker/alt status using individual opcodes
-- When a member is removed or added to a guild, a single opcode is sent to each guild member
-- This reduces network traffic considerably

Known issues:
- Visual bug only. Guild Tributes window will display a 0 for level if tribute is above max level rule setting.
- Visual bug only. Guild Mgmt Window will not display an online member within the Notes/Tribute tab if the player has 'show offline' unchecked and a guild member zones  This is resolved by selecting and de-selecting the 'Show Offline' checkbox.

There are database changes that rewrite the guild_members structure and DATA, so please only use in a testing environment.  There is also a sql required to update the guild tribute items.  This is not within source and is posted here.  It creates several items using id 150000 and above so if you have items in that range, use caution.  Suggest a test db as well, or a backup before hand.
[GuildTributes_2023_09_13_Complete.zip](https://github.com/EQEmu/Server/files/13404458/GuildTributes_2023_09_13_Complete.zip)

You will also notice a few #guild test commands.  These were/are for my testing and will be removed once the PR is finalized.
There are also a few updates to the world api for testing purposes.  These too will be removed.
And finally, the lua protocol opcode item is also fixed, again for my testing purposes.

This code has been working in a few environments for several months, therefore the functionality has been tested.  That said, please post any issues and I will look to resolve them.  Appreciate any testing that you can do.

For the devs, thanks in advance for taking a look at this and I hope it helps the community.

Neckkola.

fryguy503 commented:
FWIW - I have been running this code for months? now and its current state is very stable and the functionality is much appreciated by players.

Thank you so much Neckkola!

noudess commented:
Any chance you know how to fix mail to guild?  It dies after something like 50 members.

neckkola commented:
@noudess I will certainly take a look a see.

neckkola commented:
@noudess.  Take a look at https://github.com/EQEmu/Server/pull/3724.  Be appreciative if you can test and let me know.

Kinglykrab commented on common/item_instance.cpp:
```diff
@@ -1791,6 +1791,16 @@ int EQ::ItemInstance::RemoveTaskDeliveredItems()
 	return count;
 }
 
+int EQ::ItemInstance::GetItemGuildFavor() const
+{
+	int total = 0;
+	const auto item = GetItem();
+	if (item) {
+		return total = item->GuildFavor;
+	}
+	return 0;
+}
+
```
`uint32` maybe?

```cpp
uint32 EQ::ItemInstance::GetItemGuildFavor() const
{
	const auto item = GetItem();
	if (item) {
		return item->GuildFavor;
	}

	return 0;
}
```

neckkola replied:
Updated

Kinglykrab commented on common/item_instance.h:
```diff
@@ -299,6 +299,7 @@ namespace EQ
 		int GetItemHeroicDR(bool augments = false) const;
 		int GetItemHeroicCorrup(bool augments = false) const;
 		int GetItemHaste(bool augments = false) const;
+		int GetItemGuildFavor() const;
```
`uint32`?

neckkola replied:
Updated

Kinglykrab commented on common/patches/titanium.cpp:
```diff
@@ -664,6 +664,57 @@ namespace Titanium
+
+		//Translate older ranks to new values* /
+		switch (emu->Rank) {
+		case 8: case 7: case 6: case 5: case 4: { eq->Rank = 0; break; }	// GUILD_MEMBER  0
```
Use constants?

neckkola replied:
Updated

Kinglykrab commented on common/patches/titanium.cpp:
```diff
@@ -664,6 +664,57 @@ namespace Titanium
+			case 8: case 7: case 6: case 5: case 4: { eq->parameter = 0; break; }	// GUILD_MEMBER	 0
+			case 3: case 2: { eq->parameter = 1; break; }							// GUILD_OFFICER 1
+			case 1: { eq->parameter = 2; break; }									// GUILD_LEADER	 2
+			default: { break; }
```
Use constants?

neckkola replied:
Updated

Kinglykrab commented on common/patches/titanium.cpp:
```diff
@@ -727,12 +778,18 @@ namespace Titanium
-				PutFieldN(rank);
+//				PutFieldN(rank);
+				switch (emu_e->rank) {
+				case 8: case 7: case 6: case 5: case 4: { e->rank = htonl(0); break; }  // GUILD_MEMBER	0
+				case 3: case 2: { e->rank = htonl(1); break; }							// GUILD_OFFICER 1
+				case 1: { e->rank = htonl(2); break; }									// GUILD_LEADER	2
```
Use constants?

neckkola replied:
Updated

Kinglykrab commented on common/patches/titanium.cpp:
```diff
@@ -1864,7 +2008,13 @@ namespace Titanium
 			eq->petOwnerId = emu->petOwnerId;
-			eq->guildrank = emu->guildrank;
+			switch (emu->guildrank) {
+			case 8: case 7: case 6: case 5: case 4: { eq->guildrank = 0; break; }  // GUILD_MEMBER	0
+			case 3: case 2: { eq->guildrank = 1; break; }       				   // GUILD_OFFICER 1
+			case 1: { eq->guildrank = 2; break; }								   // GUILD_LEADER	2
```
Use constants?

neckkola replied:
Updated

Kinglykrab commented on common/patches/uf.cpp:
```diff
@@ -1134,6 +1134,12 @@ namespace UF
 				PutFieldN(class_);
+				//Translate older ranks to new values* /
+				switch (emu_e->rank) {
+				case 8: case 7: case 6: case 5: case 4: { emu_e->rank = 0; break; }
+				case 3: case 2: { emu_e->rank = 1; break; }  // GUILD_OFFICER	1
+				case 1: { emu_e->rank = 2; break; }  // GUILD_LEADER	2
```
Use constants?

neckkola replied:
Updated

Kinglykrab commented on common/patches/uf.cpp:
```diff
@@ -1829,6 +1867,12 @@ namespace UF
 		OUT(gm);
+		//Translate older ranks to new values* /
+		switch (emu->guildrank) {
+		case 8: case 7: case 6: case 5: case 4: { emu->guildrank = 0; break; }
+		case 3: case 2: { emu->guildrank = 1; break; }  // GUILD_OFFICER	1
+		case 1: { emu->guildrank = 2; break; }  // GUILD_LEADER	2
```
Use constants?

neckkola replied:
Updated

Kinglykrab commented on common/patches/uf.cpp:
```diff
@@ -2332,29 +2376,53 @@ namespace UF
+			//Translate new ranks to old values* /
+			switch (emu->parameter)
+			{
+			case 8: case 7: case 6: case 5: case 4: { eq->parameter = 0; break; }	// GUILD_MEMBER	 0
+			case 3: case 2: { eq->parameter = 1; break; }							// GUILD_OFFICER 1
+			case 1: { eq->parameter = 2; break; }									// GUILD_LEADER	 2
```
Use constants?

neckkola replied:
Updated

Kinglykrab commented on common/patches/uf.cpp:
```diff
@@ -2332,29 +2376,53 @@ namespace UF
+		
+		//Translate older ranks to new values* /
+		switch (emu->Rank) {
+		case 8: case 7: case 6: case 5: case 4: { eq->Rank = 0; break; }	// GUILD_MEMBER  0
+		case 3: case 2: { eq->Rank = 1; break; }							// GUILD_OFFICER 1
+		case 1: { eq->Rank = 2; break; }									// GUILD_LEADER	 2
```
Use constants?

neckkola replied:
Updated

Kinglykrab commented on common/patches/uf.cpp:
```diff
@@ -2945,6 +3013,12 @@ namespace UF
 				VARSTRUCT_ENCODE_TYPE(uint32, Buffer, emu->guildID);
+				//Translate older ranks to new values* /
+				switch (emu->guildrank) {
+				case 8: case 7: case 6: case 5: case 4: { emu->guildrank = 0; break; }
+				case 3: case 2: { emu->guildrank = 1; break; }  // GUILD_OFFICER	1
+				case 1: { emu->guildrank = 2; break; }  // GUILD_LEADER	2
```
Use constants?

neckkola replied:
Updated

Kinglykrab commented on common/patches/uf.cpp:
```diff
@@ -3577,6 +3651,34 @@ namespace UF
+
+		memcpy(emu->name, eq->name, sizeof(emu->name));
+		memcpy(emu->target, eq->target, sizeof(emu->target));
+		emu->rank = 5;
```
Use constant?

neckkola replied:
Updated

Kinglykrab commented on common/patches/titanium.cpp:
```diff
@@ -2205,6 +2355,34 @@ namespace Titanium
+
+		memcpy(emu->name, eq->name, sizeof(emu->name));
+		memcpy(emu->target, eq->target, sizeof(emu->target));
+		emu->rank = 5;
```
Use constant?

neckkola replied:
Updated

Kinglykrab commented on common/guild_base.cpp:
```diff
@@ -39,347 +88,310 @@ BaseGuildManager::~BaseGuildManager() {
+	{
+		std::vector<BaseGuildRanksRepository::GuildRanks> out;
+		BaseGuildRanksRepository::GuildRanks gr;
+		for (int i = 1; i <= GUILD_MAX_RANK; i++) {
```
Use constant?

neckkola replied:
Updated

Kinglykrab commented on world/eqemu_api_world_data_service.cpp:
```diff
@@ -270,8 +272,64 @@ void EQEmuApiWorldDataService::get(Json::Value &r, const std::vector<std::string
 	if (m == "get_reload_types") {
 		getReloadTypes(r);
-	}
+	}				
```
Weird whitespace here.

neckkola replied:
I don't think the white space is weird.  Looks quite normal really.  Just seems to be oddly located.   Ok, so that is my attempt at white space humor.  haha

Updated.

Kinglykrab replied:
> I don't think the white space is weird. Looks quite normal really. Just seems to be oddly located. Ok, so that is my attempt at white space humor. haha
> 
> Updated.

😆

Kinglykrab commented on zone/client_packet.cpp:
A lot of unnecessary refactoring here that isn't related to your guild code.

neckkola replied:
Cleaned up.  Not sure why all those changes found there way in there.

Kinglykrab commented on zone/client_packet.cpp:
```diff
@@ -8484,75 +8599,103 @@ void Client::Handle_OP_GuildStatus(const EQApplicationPacket *app)
+	switch (rank) {
+	case 1: {
 		MessageString(Chat::LightGray, LEADER_OF_X_GUILD, c->GetName(), GuildName);
-	else if (IsOfficer)
+		break;
+	}
+	case 2:
+	case 3: {
 		MessageString(Chat::LightGray, OFFICER_OF_X_GUILD, c->GetName(), GuildName);
-	else
+		break;
+	}
+	case 4:
+	case 5:
+	case 6:
+	case 7:
+	case 8: {
```
Use constants?

neckkola replied:
Updated

Kinglykrab commented on zone/guild_mgr.cpp:
```diff
@@ -168,14 +173,24 @@ uint8 *ZoneGuildManager::MakeGuildMembers(uint32 guild_id, const char *prefix_na
+		auto c = entity_list.GetClientByID(ci->char_id);
+		if (c && c->ClientVersion() < EQ::versions::ClientVersion::RoF) {
+			switch (ci->rank) {
+			case 8:case 7:case 6:case 5:case 4: { ci->rank = 0; break; }	//GUILD MEMBER 0
+			case 3:case 2: { ci->rank = 1; break; }							//GUILD OFFICER 1
+			case 1: { ci->rank = 2; break; }								//GUILD LEADER 2
```
Use constants?

neckkola replied:
Updated

Kinglykrab commented on zone/guild_mgr.cpp:
```diff
@@ -1454,3 +1694,244 @@ bool GuildBankManager::AllowedToWithdraw(uint32 GuildID, uint16 Area, uint16 Slo
+	{
+		auto outapp = new EQApplicationPacket(OP_GuildUpdateURLAndChannel, sizeof(GuildUpdateUCP));
+		GuildUpdateUCP* gucp = (GuildUpdateUCP*)outapp->pBuffer;
+		for (int i = 1; i <= 8; i++)
```
Use constants?

neckkola replied:
Updated

Kinglykrab commented on zone/guild_mgr.cpp:
```diff
@@ -1454,3 +1694,244 @@ bool GuildBankManager::AllowedToWithdraw(uint32 GuildID, uint16 Area, uint16 Slo
+		auto c = entity_list.GetClientByID(ci->char_id);
+		if (c && c->ClientVersion() < EQ::versions::ClientVersion::RoF) {
+			switch (ci->rank) {
+			case 8:case 7:case 6:case 5:case 4: { ci->rank = 0; break; }	//GUILD MEMBER 0
+			case 3:case 2: { ci->rank = 1; break; }							//GUILD OFFICER 1
+			case 1: { ci->rank = 2; break; }								//GUILD LEADER 2
```
Use constants?

neckkola replied:
Updated

Kinglykrab commented on zone/mob.cpp:
```diff
@@ -4000,20 +4000,26 @@ uint8 Mob::GetDefaultGender(uint16 in_race, uint8 in_gender) {
 }
 
-void Mob::SendAppearancePacket(uint32 type, uint32 value, bool WholeZone, bool iIgnoreSelf, Client *specific_target) {
+void Mob::								SendAppearancePacket(uint32 type, uint32 value, bool WholeZone, bool iIgnoreSelf, Client *specific_target, bool WholeZoneMinusMyGuild) {
```
???

neckkola replied:
Same here.  Like it had a party or something....

Updated

Kinglykrab commented:
Left some comments, `client_packet.cpp` seems to have had the entire file accidentally refactored.

noudess commented:
> @noudess. Take a look at #3724. Be appreciative if you can test and let me know.

Will do.

noudess commented:
> @noudess. Take a look at #3724. Be appreciative if you can test and let me know.

Tested and works fine here!!

Fixed: https://github.com/EQEmu/Server/issues/577 (open since 2016)!!!

Thanks!!

CoreAutomation-JD commented:
Seems like fixed item ids would be a real problem, why not check for max id then generate from there?

Akkadius commented on world/eqemu_api_world_data_service.cpp:
```diff
@@ -270,8 +272,64 @@ void EQEmuApiWorldDataService::get(Json::Value &r, const std::vector<std::string
+		row["Functions"][i]["perm_value"]	= guild->functions[i].perm_value;
+	}
+
+	row["Tribute"]["Favor"]			 = guild->tribute.favor;
```
These should all be lowercase and `snake_case`

Akkadius replied:
```
row["tribute"]["favor"]
```

neckkola replied:
updated

Akkadius commented on world/wguild_mgr.cpp:
```diff
@@ -133,15 +141,341 @@ void WorldGuildManager::ProcessZonePacket(ServerPacket *pack) {
+
+	tribute_list.clear();
+
+	const std::string query = "SELECT id, name, descr, unknown, isguild FROM tributes";
```
Repository?

neckkola replied:
Updated

hgtw commented on common/guild_base.h:
```diff
@@ -5,6 +5,43 @@
+#include "timer.h"
+#include "../common/repositories/guild_members_repository.h"
+
+#define GOUT(x) out.x = in->x;
```
If this macro is going to be used only inside `guild_base.cpp` it should just be placed at the top of that file to avoid leaking the macro to anything that includes this header.

Also not sure we should be using macros just for assignment like this outside of packet encoder stuff but maybe it simplified something

neckkola replied:
Created the routing to copy the info and therefore no longer requiring the macro.  Removed

hgtw commented on common/guild_base.cpp:
```diff
@@ -39,347 +88,310 @@ BaseGuildManager::~BaseGuildManager() {
+		BaseGuildsRepository::Guilds out;
+		out.id = guild_id;
+		out.favor = in->tribute.favor;
+		GOUT(channel);
```
It looks like this is a common pattern copying `GuildInfo*` to `BaseGuildsRepository::Guilds`. Maybe a private `BaseGuildManager` function to return a `BaseGuildsRepository::Guilds` from a  `GuildInfo*` argument would be better to avoid duplicating the logic everywhere this is done. Then the `GOUT` macro could also be removed since the assignments would only occur in one place.

e.g.
`BaseGuildsRepository::Guilds out = CreateGuildsRepoEntry(guild_id, in);`
(maybe with a more clever function name)

neckkola replied:
Thanks for the discussion.  Made a function to look after this and found an oversight on my part in the process.  More evidence to have a single entity to reduce human error.

Akkadius commented on common/patches/titanium.cpp:
```diff
@@ -867,11 +956,60 @@ namespace Titanium
+		};
+
+		std::vector<bit_mask_conversion> bit_mask = {
+			{2, 256},
```
I'd inline this for readability

```{.ti_mask = 2, rof2_mask = 256},```

Akkadius replied:
I did it for you

neckkola replied:
Thank you.

Akkadius commented on common/repositories/guild_members_repository.h:
```diff
@@ -0,0 +1,193 @@
+				"JOIN character_data cd ON cd.id = gm.char_id "
+				"WHERE ISNULL(cd.deleted_at) AND gm.tribute_enable = 1 AND gm.guild_id = '{}';",
+				guild_id
+				)
```
Some weird formatting

neckkola replied:
Cleaned up.

Akkadius commented on common/repositories/guild_members_repository.h:
```diff
@@ -0,0 +1,193 @@
+			e.char_id		 = static_cast<int32_t>(atoi(row[0]));
+			e.guild_id		 = static_cast<uint32_t>(strtoul(row[1], nullptr, 10));
+			e.tribute_enable = static_cast<uint8_t>(strtoul(row[2], nullptr, 10));
+			e.char_name		 = row[3] ? row[3] : "";
```
Indenting is not consistent

neckkola replied:
Cleaned up.  In the process of these weird formatting realized that there are some editor functions on my end to update.  :-)

Akkadius commented on common/repositories/guilds_repository.h:
```diff
@@ -5,46 +5,91 @@
+				TableName(),
+				favor,
+				guild_id
+				)
```
Indenting

neckkola replied:
Cleaned up

Akkadius commented on queryserv/lfguild.cpp:
```diff
@@ -10,27 +10,27 @@
 }
 
-GuildLookingForPlayers::GuildLookingForPlayers(char *Name, char *Comments, uint32 FromLevel, uint32 ToLevel, uint32 Classes, uint32 AACount, uint32 Timezone, uint32 TimePosted)
+GuildLookingForPlayers::GuildLookingForPlayers(char * g_Name, char * g_Comments, uint32 g_FromLevel, uint32 g_ToLevel, uint32 g_Classes, uint32 g_AACount, uint32 g_Timezone, uint32 g_TimePosted)
```
Why are all of these prefixed with `g_`? That usually implies global variable

If anything make the `GuildLookingForPlayers` member variables private and prefix them with `m_` (member variable) so you don't need to prefix input parameters to the function.

eg `m_name = std::string(name)`

neckkola replied:
Updated this and the previous function to be consistent.

Akkadius commented on common/repositories/guild_members_repository.h:
```diff
@@ -0,0 +1,193 @@
+		return (results.Success() ? results.RowsAffected() : 0);
+	}
+
+	static int UpdateOnline(Database& db, uint32 char_id, bool status)
```
Your base repository was missing the `online` field so I regenerated it for you and pushed it up

Instead of doing raw queries in custom repository methods, you can lean into the core functionality of repositories here.

```cpp
static int UpdateOnline(Database& db, uint32 char_id, bool status)
{
	const auto guild_members = GetWhere(db, fmt::format("char_id = '{}'", char_id));
	if (guild_members.empty()) {
		return 0;
	}

	auto m = guild_members[0];

	m.online = status ? 1 : 0;

	return UpdateOne(db, m);
}
```

Needs to be tested of course but should work. You are querying and use an active record pattern to manipulate the struct and then pass the data model back to an update method and the repository method takes care of doing the update for you. You're doing a lot of the work the base repositories are already doing

Akkadius replied:
Same thing goes for other methods - you can re-use the same pattern and lean into the core repository methods

neckkola replied:
Refactored all to better use the repos.  Thanks for the advice.

Akkadius commented on common/guild_base.h:
```diff
@@ -105,31 +153,36 @@ class BaseGuildManager
 		virtual void SendRankUpdate(uint32 CharID) = 0;
 		virtual void SendGuildDelete(uint32 guild_id) = 0;
 
-		uint32	DBCreateGuild(const char* name, uint32 leader_char_id);
```
I know you're following existing patterns here but these don't need `DB` as a prefix, just describe what they do and then maybe we call a repository method underneath

neckkola replied:
Updated them all and propagated throughout.

Akkadius commented on common/repositories/guild_members_repository.h:
```diff
@@ -0,0 +1,193 @@
+		return (results.Success() ? results.RowsAffected() : 0);
+	}
+
+	static std::vector<GuildMembersWithTributeOn> GetMembersWithTributeOn(Database& db, uint32 guild_id)
```
Love that you're returning a custom response object for more specifically formatting joined data

neckkola replied:
Thanks.

Akkadius commented on common/repositories/guild_permissions_repository.h:
```diff
@@ -0,0 +1,107 @@
+
+	// Custom extended repository methods here
+
+	static int UpdateOne(
```
There's already an `UpdateOne` in the base repository - is there a reason we duplicated it here ?

neckkola replied:
Refactored all to better use the repos.

Akkadius commented on common/repositories/guild_permissions_repository.h:
```diff
@@ -0,0 +1,107 @@
+		return (results.Success() ? results.RowsAffected() : 0);
+	}
+
+	static int ReplaceMany(
```
Did we need a `ReplaceMany` when we have `InsertMany` ? You can delete the data before re-inserting it just as well. That's essentially what `REPLACE` is doing under the hood anyhow.

Live example

```cpp
CharacterSpellsRepository::DeleteWhere(database, fmt::format("id = {}", CharacterID()));

if (!character_spells.empty()) {
	CharacterSpellsRepository::InsertMany(database, character_spells);
}
```

If you search the codebase for `::InsertMany(` you'll see many occurrences of this today.

neckkola replied:
Refactored all to better use the repos.

Akkadius commented on common/repositories/guild_tributes_repository.h:
```diff
@@ -0,0 +1,119 @@
+		);
+	}
+	
+	static GuildTributes ReplaceOne(
```
You know you can use `UpdateOne` in place of what this is doing as long as you have a primary key to anchor to. I gave an example in another comment

neckkola replied:
Refactored all to better use the repos.

Akkadius commented on common/repositories/guilds_repository.h:
```diff
@@ -5,46 +5,91 @@
 
+	 // Custom extended repository methods here
+
+	static int ReplaceOne(
```
Same thing - did you need `ReplaceOne` when `UpdateOne` will do a lot of what you're trying to do here

neckkola replied:
Refactored all to better use the repos.

Akkadius commented on world/clientlist.cpp:
```diff
@@ -1667,3 +1673,130 @@ void ClientList::SendCharacterMessageID(ClientListEntry* character,
+	return guild_members;
+}
+
+std::map<uint32, ClientListEntry*> ClientList::GetGuildClientList(uint32 guild_id) {
```
It appears this function is not used

neckkola replied:
Cleaned up

Akkadius commented on world/clientlist.cpp:
```diff
@@ -1667,3 +1673,130 @@ void ClientList::SendCharacterMessageID(ClientListEntry* character,
+	return guild_members;
+}
+
+//Client* ClientList::GetClient(uint32 char_id) 
```
Remove?

neckkola replied:
Cleaned up

Akkadius commented on world/eqemu_api_world_data_service.cpp:
```diff
@@ -270,8 +272,64 @@ void EQEmuApiWorldDataService::get(Json::Value &r, const std::vector<std::string
+	}
+
+	for (int i = 1; i <= 30; i++) {
+		row["Functions"][i]["db_id"]		= guild->functions[i].id;
```
lower case snake_case keys

neckkola replied:
updated

Akkadius commented on world/eqemu_api_world_data_service.cpp:
```diff
@@ -270,8 +272,64 @@ void EQEmuApiWorldDataService::get(Json::Value &r, const std::vector<std::string
+	row["Tribute"]["ID1 Tier"]		 = guild->tribute.id_1_tier;
+	row["Tribute"]["ID2"]			 = guild->tribute.id_2;
+	row["Tribute"]["ID2 Tier"]		 = guild->tribute.id_2_tier;
+	row["Tribute"]["Time Remaining"] = guild->tribute.time_remaining;
```
Alignment

neckkola replied:
updated

Akkadius commented on world/wguild_mgr.cpp:
```diff
@@ -133,15 +141,341 @@ void WorldGuildManager::ProcessZonePacket(ServerPacket *pack) {
+			guild_mgr.DBSetGuildFavor(g.first, g.second->tribute.favor);
+			guild_mgr.DBSetTributeTimeRemaining(g.first, RuleI(Guild, TributeTime));
+
+			LogGuilds("Timer reset.  Do tribute work - Points, send timer restart to zones, etc.");
```
Is this a helpful log message?

neckkola replied:
Was from my testing.  Removed.

Akkadius commented on world/wguild_mgr.cpp:
```diff
@@ -133,15 +141,341 @@ void WorldGuildManager::ProcessZonePacket(ServerPacket *pack) {
+			LogGuilds("Timer reset.  Do tribute work - Points, send timer restart to zones, etc.");
+
+			SendGuildTributeFavorAndTimer(g.first, g.second->tribute.favor, g.second->tribute.timer.GetRemainingTime());
+			//ServerPacket* sp = new ServerPacket(ServerOP_GuildTributeFavAndTimer, sizeof(GuildTributeFavorTimer_Struct));
+			//GuildTributeFavorTimer_Struct* data = (GuildTributeFavorTimer_Struct*)sp->pBuffer;
+			//data->guild_id = g.first;
+			//data->guild_favor = g.second->tribute.favor;
+			//data->tribute_timer = g.second->tribute.time_remaining;
+			//data->trophy_timer = 0;
+			//
+			//for (auto const& z : zoneserver_list.getZoneServerList()) {
+			//	auto r = z.get();
+			//	if (r->GetZoneID() > 0) {
+			//		r->SendPacket(sp);
+			//	}
+			//}
+			//safe_delete(sp);
```
?

neckkola replied:
Sorry, left over.  Cleaned up.

Akkadius commented on world/wguild_mgr.cpp:
```diff
@@ -133,15 +141,341 @@ void WorldGuildManager::ProcessZonePacket(ServerPacket *pack) {
+
+	LogInfo("Loaded [{}] tributes", Strings::Commify(results.RowCount()));
+
+	const std::string query2 = "SELECT tribute_id, level, cost, item_id FROM tribute_levels ORDER BY tribute_id, level";
```
Repository call

neckkola replied:
Updated as noted below.

Akkadius commented on world/wguild_mgr.cpp:
```diff
@@ -133,15 +141,341 @@ void WorldGuildManager::ProcessZonePacket(ServerPacket *pack) {
+	return total_cost;
+}
+
+bool WorldGuildManager::LoadTributes() {
```
Put together a compiling example of what this function can look like with repository calls. Code is simpler and easier to read. It does need to be fully tested.

```cpp
bool WorldGuildManager::LoadTributes()
{
	TributeData td;
	memset(&td.tiers, 0, sizeof(td.tiers));
	td.tier_count = 0;

	tribute_list.clear();

	auto tributes = TributesRepository::All(*m_db);

	for (auto const &t: tributes) {
		td.name        = t.name;
		td.description = t.descr;
		td.unknown     = t.unknown;
		td.is_guild    = t.isguild == 0 ? false : true;

		tribute_list[t.id] = td;
	}

	LogInfo("Loaded [{}] tributes", Strings::Commify(tributes.size()));

	auto tribute_levels = TributeLevelsRepository::GetWhere(*m_db, "TRUE ORDER BY tribute_id, level");

	for (auto const &t: tribute_levels) {
		uint32 id = t.tribute_id;

		if (tribute_list.count(id) != 1) {
			LogError("Error in LoadTributes: unknown tribute [{}] in tribute_levels", (unsigned long) id);
			continue;
		}

		TributeData &cur = tribute_list[id];
		if (cur.tier_count >= MAX_TRIBUTE_TIERS) {
			LogError("Error in LoadTributes: on tribute [{}] more tiers defined than permitted", (unsigned long) id);
			continue;
		}

		TributeLevel_Struct &s = cur.tiers[cur.tier_count];

		s.level           = t.level;
		s.cost            = t.cost;
		s.tribute_item_id = t.item_id;

		cur.tier_count++;
	}

	LogInfo("Loaded [{}] tribute levels", Strings::Commify(tribute_levels.size()));

	return true;
}
```

neckkola replied:
Updated for next commit

Akkadius commented on world/wguild_mgr.cpp:
```diff
@@ -133,15 +141,341 @@ void WorldGuildManager::ProcessZonePacket(ServerPacket *pack) {
+	}
+
+	where_filter = fmt::format("guild_id = '{}'", guild_id);
+	auto g_permissions = BaseGuildPermissionsRepository::GetWhere(*m_db, where_filter);
```
`g_` implies global I'd just use `permissions`

neckkola replied:
I used g_ to represent guild_tributes to not be confused with tributes, which is used for player tributes.

Akkadius replied:
It's well understood across cpp projects that `g_` prefixed variables are global scoped. I'd prefer we remove them altogether and find another way to identify member variables such as making them private and prefixing them with `m_` 

https://stackoverflow.com/questions/22736150/global-variables-what-are-they-exactly

neckkola replied:
Sounds good.

neckkola replied:
Updated to be guild_ in the next commit.

Akkadius commented on world/zoneserver.cpp:
```diff
@@ -978,7 +980,14 @@ void ZoneServer::HandleMessage(uint16 opcode, const EQ::Net::Packet &p) {
 		case ServerOP_GuildCharRefresh:
 		case ServerOP_GuildMemberUpdate:
-		case ServerOP_RefreshGuild: {
+		case ServerOP_GuildPermissionUpdate:
+		case ServerOP_GuildRankNameChange:
+		case ServerOP_RefreshGuild: 
+		case ServerOP_GuildMemberLevelUpdate: 
+		case ServerOP_GuildMemberPublicNote:
+		case ServerOP_GuildMemberRemove:
+		case ServerOP_GuildMemberAdd:
+		{
```
Great!

neckkola replied:
Gazoo!

Akkadius replied:
LOL

Akkadius commented on zone/guild.cpp:
```diff
@@ -234,8 +246,11 @@ void Client::RefreshGuildInfo()
 	guild_id = info.guild_id;
-	GuildBanker = info.banker || guild_mgr.IsGuildLeader(GuildID(), CharacterID());
-
+	GuildBanker = info.banker ||
```
What is happening here ?

neckkola replied:
This could is my attempt for backwards compatibility with TI/UF.  These clients have a banker flag, while RoF uses the permissions system.  Therefore, I have set the GuildBanker Flag to be on, if the flag is set, the player is the guild leader, a guild leader using the TI/UF client sets the banker flag, or if the Client is an RoF client.

Then, in the banker system, it checks for the flag and the the RoF permission.

Akkadius commented on zone/guild_mgr.cpp:
```diff
@@ -334,40 +349,40 @@ void ZoneGuildManager::DescribeGuild(Client *c, uint32 guild_id) const {
-			can_war_peace
-		);
-	}
+					
+	//for (uint8 guild_rank = 0; guild_rank <= GUILD_MAX_RANK; guild_rank++) {
+	//	auto can_hear_guild_chat = info->rank_s[guild_rank].permissions[GUILD_HEAR] ? "<c \"#00FF00\">Y</c>" : "<c \"#F62217\">N</c>";
+	//	auto can_speak_guild_chat = info->ranks[guild_rank].permissions[GUILD_SPEAK] ? "<c \"#00FF00\">Y</c>" : "<c \"#F62217\">N</c>";
+	//	auto can_invite = info->ranks[guild_rank].permissions[GUILD_INVITE] ? "<c \"#00FF00\">Y</c>" : "<c \"#F62217\">N</c>";
+	//	auto can_remove = info->ranks[guild_rank].permissions[GUILD_REMOVE] ? "<c \"#00FF00\">Y</c>" : "<c \"#F62217\">N</c>";
+	//	auto can_promote = info->ranks[guild_rank].permissions[GUILD_PROMOTE] ? "<c \"#00FF00\">Y</c>" : "<c \"#F62217\">N</c>";
+	//	auto can_demote = info->ranks[guild_rank].permissions[GUILD_DEMOTE] ? "<c \"#00FF00\">Y</c>" : "<c \"#F62217\">N</c>";
+	//	auto can_set_motd = info->ranks[guild_rank].permissions[GUILD_MOTD] ? "<c \"#00FF00\">Y</c>" : "<c \"#F62217\">N</c>";
+	//	auto can_war_peace = info->ranks[guild_rank].permissions[GUILD_WARPEACE] ? "<c \"#00FF00\">Y</c>" : "<c \"#F62217\">N</c>";
+	//	popup_text += fmt::format(
+	//		"<tr>"
+	//		"<td>{} ({})</td>"
+	//		"<td>{}</td>"
+	//		"<td>{}</td>"
+	//		"<td>{}</td>"
+	//		"<td>{}</td>"
+	//		"<td>{}</td>"
+	//		"<td>{}</td>"
+	//		"<td>{}</td>"
+	//		"<td>{}</td>"
+	//		"</tr>",
+	//		!info->ranks[guild_rank].name.empty() ? info->ranks[guild_rank].name : "Nameless",
+	//		guild_rank,
+	//		can_demote,
+	//		can_hear_guild_chat,
+	//		can_invite,
+	//		can_promote,
+	//		can_remove,
+	//		can_set_motd,
+	//		can_speak_guild_chat,
+	//		can_war_peace
+	//	);
+	//}
```
?

neckkola replied:
The old guild system had a popup to display info that appeared to mirror the RoF permission system.  I should remove this as the permission system displays the data now.

Akkadius commented on zone/guild_mgr.cpp:
```diff
@@ -396,69 +411,85 @@ bool ZoneGuildManager::VerifyAndClearInvite(uint32 char_id, uint32 guild_id, uin
-				c->SendGuildMembers();	//only need to update this player's list (trying to clear it)
-		} else {
+		if (s->guild_id == GUILD_NONE) {
+			if (c != nullptr) {
```
`if (!c)` does the same thing

neckkola replied:
Updated several occurrences of this in next commit.

Akkadius commented on zone/guild_mgr.cpp:
```diff
@@ -473,86 +504,150 @@ void ZoneGuildManager::ProcessWorldPacket(ServerPacket *pack) {
+		auto IsActionABankAction = [&](GuildAction action) -> bool {
+			std::vector<GuildAction> Guild_Bank_Actions =
+			{
+			GUILD_ACTION_BANK_DEPOSIT_ITEMS,
```
Formatting

neckkola replied:
Updated in next commit

neckkola replied:
Actually had this lambda in a few spots, so converted to a regular function.

Akkadius commented on zone/guild_mgr.cpp:
```diff
@@ -473,86 +504,150 @@ void ZoneGuildManager::ProcessWorldPacket(ServerPacket *pack) {
+			else if (!guild_bank_status && !sgrus->no_update) {
+				entity_list.GuildSetPreRoFBankerFlag(sgrus->GuildID, sgrus->Rank, false);
+			}
+			//auto client_list = entity_list.GetClientList();
+			//for (auto const& c : client_list) {
+			//	if (c.second->GuildID() == sgrus->GuildID) {
+			//		CharGuildInfo cgi;
+			//		guild_mgr.GetCharInfo(c.second->CharacterID(), cgi);
+			//		if (guild_bank_status && !cgi.banker) {
+			//			guild_mgr.SetBankerFlag(c.second->CharacterID(), true);
+			//			gsrs->Banker = 1 + (cgi.alt ? 2 : 0);
+			//		}
+			//		else if (!guild_bank_status && cgi.banker) {
+			//			guild_mgr.SetBankerFlag(c.second->CharacterID(), false);
+			//			gsrs->Banker = 0 + (cgi.alt ? 2 : 0);
+			//		}
+			//	}
+			//	c.second->QueuePacket(outapp);
+			//}
+			//safe_delete(outapp);
+
+//		Client* c = entity_list.GetClientByName(sgrus->MemberName);
+//		if (c) {
+//			c->SendAppearancePacket(AT_GuildRank, sgrus->Rank, false);
+//		}
+//		entity_list.SendAllGuildTitleDisplay(sgrus->GuildID);
```
Remove ?

neckkola replied:
Removed

Akkadius commented on zone/guild_mgr.cpp:
```diff
@@ -561,10 +656,10 @@ void ZoneGuildManager::ProcessWorldPacket(ServerPacket *pack) {
 			char Name[64];
 			gmus->LastSeen = time(nullptr);
-			gmus->InstanceID = 0;
+			gmus->InstanceID = 1;
```
Why is this flipped ?

neckkola replied:
I traced this to allow for correct display within the Guild Mgmt Window for online members

Akkadius commented on zone/guild_mgr.cpp:
```diff
@@ -615,10 +710,155 @@ void ZoneGuildManager::ProcessWorldPacket(ServerPacket *pack) {
+				entity_list.QueueClientsGuild(outapp, s->guild_id);
+				safe_delete(outapp);
+
+				//for (auto const& gm : guild_mgr.GetGuildMembers(s->guild_id)) {
```
Remove ?

neckkola replied:
Removed in next commit

Akkadius commented on zone/guild_mgr.cpp:
```diff
@@ -1454,3 +1694,244 @@ bool GuildBankManager::AllowedToWithdraw(uint32 GuildID, uint16 Area, uint16 Slo
+
+void ZoneGuildManager::AddMember(uint32 guild_id, uint32 char_id, uint32 level, uint32 _class, uint32 rank, uint32 zone_id, std::string player_name)
+{
+	BaseGuildMembersRepository::GuildMembers member;
```
You can make this a single letter to condense verbosity. `member` is a complex type so `m` doesn't make it any less readable, infact makes it more readable

neckkola replied:
Updated for next commit

Akkadius commented on zone/inventory.cpp:
```diff
@@ -1098,13 +1098,24 @@ void Client::DeleteItemInInventory(int16 slot_id, int16 quantity, bool client_up
 		if(update_db)
 			database.SaveCursor(character_id, s, e);
 	}
+	//else if (slot_id >= EQ::invslot::GUILD_TRIBUTE_BEGIN && slot_id <= EQ::invslot::GUILD_TRIBUTE_END) {
```
Remove?

neckkola replied:
Updated for next commit

Akkadius commented on common/patches/titanium_structs.h:
```diff
@@ -3699,6 +3716,15 @@ struct SayLinkBodyFrame_Struct {
 
+struct GuildSetRank_Struct
+{
+	/*00*/	uint32	Unknown00;
```
`snake_case` on fields ?

neckkola replied:
Updated across all the struct changes.

Akkadius commented on common/eq_packet_structs.h:
```diff
@@ -3514,15 +3551,142 @@ struct TributeAbility_Struct {
+};
+
+struct GuildTributeDonateItemReply_Struct {
+/*000*/ uint32	Type;
```
`snake_case` on these fields

Akkadius commented on common/eq_packet_structs.h:
```diff
@@ -3496,66 +3533,193 @@ struct ZoneInSendName_Struct2 {
+};
+
+struct GuildTributeDonateItemRequest_Struct {
+/*000*/    uint32 Type;
```
In general these fields should be `snake_case` - I know previous structs are not consistent but we are trying to be consistent going forward especially during code review

neckkola replied:
Updated

Akkadius commented on common/guild_base.cpp:
```diff
@@ -27,8 +35,49 @@
 
-const char *const BaseGuildManager::GuildActionNames[_MaxGuildAction] =
-{ "HearGuildChat", "SpeakGuildChat", "Invite", "Remove", "Promote", "Demote", "Set_MOTD", "War/Peace" };
+std::vector<default_permission_struct> default_permissions = {
```
What do these values represent (the second one) ?

Akkadius replied:
Struct `DefaultPermissionStruct` vs `default_permission_struct` ?

neckkola replied:
The GuildActionNames were the old guild system equivalent for the RoF permission system.  
The default_permissions represents the defaults for guild creation.  These were mapped from Live in summer of 2023.

Kinglykrab replied:
Think he just wants `default_permission_struct` changed to `DefaultPermissionStruct` for consistency in our struct nomenclature.

Akkadius replied:
Yep @Kinglykrab

neckkola replied:
Updated

Akkadius commented on zone/gm_commands/guild.cpp:
```diff
@@ -476,9 +502,215 @@ void command_guild(Client *c, const Seperator *sep)
 			}
 		}
 	}
+	else if (is_details)
```
@Kinglykrab have suggestions for this section ?

Kinglykrab replied:
I would personally use a popup for the data type stuff.

Seems there's a lot of unnecessary reformatting here as well.

neckkola replied:
Updated the #guild info and the DescribeGuild function.

Kinglykrab commented on zone/gm_commands/guild.cpp:
```diff
@@ -476,9 +502,215 @@ void command_guild(Client *c, const Seperator *sep)
+			}
+		}
+	}
+	else if (is_test)
```
Is this a real command or just something bogus you were messing with?

Kinglykrab replied:
Seems like bogus code to me just based on it all being test stuff for making sure things are sent properly.

neckkola replied:
It is for my testing as noted in the original notes on the PR.  I added several guild commands for easy testing of functions.  I would want to remove them once things are finalized.  I could do the same for api functions.  I will take your lead.

neckkola replied:
I have reverted to the master guild.cpp in order to remove all the test commands.

neckkola commented on common/patches/titanium.cpp:
```diff
@@ -1149,7 +1287,13 @@ namespace Titanium
+		switch (emu->guildrank) {
+		case 8: case 7: case 6: case 5: case 4: { eq->guildrank = 0; break; }  // GUILD_MEMBER	0
+		case 3: case 2: { eq->guildrank = 1; break; }       				   // GUILD_OFFICER 1
+		case 1: { eq->guildrank = 2; break; }								   // GUILD_LEADER	2
```
Updated

neckkola commented on zone/mob.cpp:
```diff
@@ -4000,20 +4000,26 @@ uint8 Mob::GetDefaultGender(uint16 in_race, uint8 in_gender) {
 	appearance->type = type;
 	appearance->parameter = value;
-	if (WholeZone)
+	if (WholeZoneMinusMyGuild) {
```
Updated

neckkola commented on common/guild_base.cpp:
```diff
@@ -39,347 +88,310 @@ BaseGuildManager::~BaseGuildManager() {
 }
 
-BaseGuildManager::GuildInfo *BaseGuildManager::_CreateGuild(uint32 guild_id, const char *guild_name, uint32 leader_char_id, uint8 minstatus, const char *guild_motd, const char *motd_setter, const char *Channel, const char *URL)
+BaseGuildManager::GuildInfo* BaseGuildManager::_CreateGuild(uint32 guild_id, std::string guild_name, uint32 leader_char_id, uint8 minstatus, std::string guild_motd, std::string motd_setter, std::string Channel, std::string URL, uint32 favor)
```
Good morning.  Happy Christmas Eve!  hg, thanks for the advice.  I have updated.  Let me know if there is anything that I missed.

neckkola commented on common/guild_base.cpp:
```diff
@@ -39,347 +88,310 @@ BaseGuildManager::~BaseGuildManager() {
+	
+	LogGuilds("Found {} guilds.  Loading.....", guilds.size());
+	for (auto const& g : guilds) {
+		_CreateGuild(g.id, g.name.c_str(), g.leader, g.minstatus, g.motd.c_str(), g.motd_setter.c_str(), g.channel.c_str(), g.url.c_str(), g.favor);
```
Updated.

neckkola commented on common/guild_base.cpp:
```diff
@@ -39,347 +88,310 @@ BaseGuildManager::~BaseGuildManager() {
-	{
-		return false;
+	LogGuilds("Found guild id [{}].  Loading details.....", db_guild.id);
+	_CreateGuild(db_guild.id, db_guild.name.c_str(), db_guild.leader, db_guild.minstatus, db_guild.motd.c_str(), db_guild.motd_setter.c_str(), db_guild.channel.c_str(), db_guild.url.c_str(), db_guild.favor);
```
Updated

neckkola commented on common/database/database_update_manifest.cpp:
```diff
@@ -5030,6 +5030,56 @@ ADD COLUMN `min_time` smallint(4) NOT NULL DEFAULT 0 AFTER `condition_value_filt
+				`perm_id` INT(11) NOT NULL DEFAULT '0',
+				`guild_id` INT(11) NOT NULL DEFAULT '0',
+				`permission` INT(11) NOT NULL DEFAULT '0',
+				PRIMARY KEY (`id`) USING BTREE
```
I put them there to enforce no duplicates of permissions within a guild.

neckkola commented on common/guild_base.cpp:
```diff
@@ -39,347 +88,310 @@ BaseGuildManager::~BaseGuildManager() {
-		if(res == m_guilds.end()) {
-			LogGuilds("Found rank [{}] for non-existent guild [{}], skipping", rankn, guild_id);
-			continue;
+		where_filter = fmt::format("guild_id = '{}'", g.first);
```
update

neckkola commented:
Good evening.  I believe that I have addressed the review/comments.  Any other adjustments, please let me know.  
Once complete, I will run a few more regression tests.

Merry Christmas everyone!

Akkadius commented on zone/client_packet.cpp:
```diff
@@ -16533,3 +16644,306 @@ void Client::ReloadExpansionProfileSetting()
+	LogTribute("Received OP_Handle_OP_GuildModifyBenefits of length [{}]", app->size);
+
+	if (app->size != sizeof(GuildTributeModifyBenefits_Struct)) {
+		LogError("Invalid size on Handle_OP_GuildModifyBenefits packet");
```
return early in this if statement so we can remove the big `else {` block

Akkadius commented on zone/client_packet.cpp:
```diff
@@ -16533,3 +16644,306 @@ void Client::ReloadExpansionProfileSetting()
+		GuildTributeModifyBenefits_Struct* gmbs = (GuildTributeModifyBenefits_Struct*)outapp->pBuffer;
+
+		switch (t->command) {
+		case 0: {
```
Magic number?

neckkola replied:
Updated several of these.

Akkadius commented on zone/client_packet.cpp:
```diff
@@ -16533,3 +16644,306 @@ void Client::ReloadExpansionProfileSetting()
+	LogTribute("Received OP_Handle_OP_GuildTributeToggle of length [{}]", app->size);
+
+	if (app->size != sizeof(GuildTributeToggleReq_Struct)) {
+		LogError("Invalid size on Handle_OP_GuildModifyBenefits packet");
```
Return early here and remove the `else {`

neckkola replied:
Done

Akkadius commented on zone/client_packet.cpp:
```diff
@@ -16533,3 +16644,306 @@ void Client::ReloadExpansionProfileSetting()
+				favor,
+				GetZoneID()
+			);
+			QServ->PlayerLogEvent(Player_Log_Guild_Tribute_Plat_Donation, CharacterID(), event_desc);
```
QServ isn't used anymore. We should use the modern player event system to capture this

neckkola replied:
Sounds good.  I will update

Akkadius commented on zone/client_packet.cpp:
```diff
@@ -16533,3 +16644,306 @@ void Client::ReloadExpansionProfileSetting()
+		SendGuildTributeDonateItemReply(in, favor);
+
+		if (RuleB(QueryServ, PlayerLogGuildTributeItemDonations)) {
+			auto event_desc = fmt::format("Guild Item Donation :: {} donated item id {} for {} favor in zone {}\.",
```
QServ isn't used anymore. We should use the modern player event system to capture this

Akkadius commented on zone/client_packet.cpp:
```diff
@@ -16533,3 +16644,306 @@ void Client::ReloadExpansionProfileSetting()
+	LogTribute("Received OP_GuildSelectTribute of length [{}]", app->size);
+
+	if (app->size != sizeof(GuildTributeSelectReq_Struct))
+		LogError("Invalid size on OP_GuildSelectTribute packet");
```
Return early and remove the `else {`

neckkola replied:
Updated

Akkadius commented on zone/client_packet.cpp:
```diff
@@ -16533,3 +16644,306 @@ void Client::ReloadExpansionProfileSetting()
+
+void Client::Handle_OP_GuildTributeSelect(const EQApplicationPacket* app)
+{
+	LogTribute("Received OP_GuildSelectTribute of length [{}]", app->size);
```
We get this logging for free with our existing packet logging, can remove

neckkola replied:
Removed.

Akkadius commented on zone/client_packet.cpp:
```diff
@@ -16533,3 +16644,306 @@ void Client::ReloadExpansionProfileSetting()
+
+void Client::Handle_OP_GuildTributeModifyBenefits(const EQApplicationPacket* app)
+{
+	LogTribute("Received OP_Handle_OP_GuildModifyBenefits of length [{}]", app->size);
```
We get this logging for free with our existing packet logging, can remove

Akkadius commented on zone/client_packet.cpp:
```diff
@@ -16533,3 +16644,306 @@ void Client::ReloadExpansionProfileSetting()
+
+void Client::Handle_OP_GuildTributeOptInOut(const EQApplicationPacket* app)
+{
+	LogTribute("Received OP_Handle_OP_GuildTributeOptInOut of length [{}]", app->size);
```
We get this logging for free with our existing packet logging, can remove

neckkola replied:
Removed.

Akkadius commented on zone/client_packet.cpp:
```diff
@@ -16533,3 +16644,306 @@ void Client::ReloadExpansionProfileSetting()
+
+void Client::Handle_OP_GuildTributeSaveActiveTributes(const EQApplicationPacket* app)
+{
+	LogTribute("Received OP_Handle_OP_GuildTributeSaveActive of length [{}]", app->size);
```
We get this logging for free with our existing packet logging, can remove

neckkola replied:
Removed.

Akkadius commented on zone/client_packet.cpp:
```diff
@@ -16533,3 +16644,306 @@ void Client::ReloadExpansionProfileSetting()
+
+void Client::Handle_OP_GuildTributeDonateItem(const EQApplicationPacket* app)
+{
+	LogTribute("Received OP_Handle_OP_GuildTributeDonateItem of length [{}]", app->size);
```
We get this logging for free with our existing packet logging, can remove

neckkola replied:
Removed.

Akkadius commented on zone/client_packet.cpp:
```diff
@@ -16533,3 +16644,306 @@ void Client::ReloadExpansionProfileSetting()
+
+void Client::Handle_OP_GuildTributeDonatePlat(const EQApplicationPacket* app)
+{
+	LogTribute("Received OP_Handle_OP_GuildTributeToggle of length [{}]", app->size);
```
We get this logging for free with our existing packet logging, can remove

neckkola replied:
Removed.

Akkadius commented on zone/guild.cpp:
```diff
@@ -121,28 +121,49 @@ void Client::SendGuildRanks()
+	if (IsInAGuild() && (ClientVersion() >= EQ::versions::ClientVersion::RoF))
+	{
+		auto guild = guild_mgr.GetGuildByGuildID(GuildID());
+		for (int i = 1; i <= 8; i++)
```
Constants

neckkola replied:
Yeah, this 8.  :-)

Akkadius commented on zone/guild.cpp:
```diff
@@ -121,28 +121,49 @@ void Client::SendGuildRanks()
+
+			gucp->payload.rank_name.rank = i;
+			strcpy(gucp->payload.rank_name.rank_name, guild->rank_names[i].c_str());
+			gucp->action = 4;
```
Constant, what is action 4 ?

neckkola replied:
Updated this and also the 8 a few lines above.

Akkadius commented on zone/client.h:
```diff
@@ -725,6 +733,23 @@ class Client : public Mob
+	void RequestGuildActiveTributes(uint32 guild_id);
+	void RequestGuildFavorAndTimer(uint32 guild_id);
+	void SendGuildMembersList();
+	void SendGuildMemberAdd(uint32 guild_id, uint32 level, uint32 _class, uint32 rank, uint32 spirit, uint32 zone_id, std::string player_name);
```
![image](https://github.com/EQEmu/Server/assets/3319450/d53df3c4-8fc7-4155-9eaf-0338ecd240cb)

All of these (in gray) are un-used - what's the story here ?

Akkadius replied:
![image](https://github.com/EQEmu/Server/assets/3319450/6d8ee883-5c12-49bb-8f03-183983f973b5)

neckkola replied:
These are the result of adding single entry opcodes for client though I did not use all of them as several are entity based.  They are present in case they are required.  I would suggest that they will be used as I work through the items noted below, for single opcode changes instead of sending the guild member list each time.

Akkadius commented on zone/client_packet.cpp:
```diff
@@ -16533,3 +16644,306 @@ void Client::ReloadExpansionProfileSetting()
+			safe_delete(sp);
+			break;
+		}
+		case 1: {
```
Magic number?

neckkola replied:
Updated here and a few others

Akkadius commented on zone/entity.h:
```diff
@@ -240,11 +240,24 @@ class EntityList
+	void SendGuildMembers(uint32 guild_id);
+	void SendGuildMembersList(uint32 guild_id);
+	void SendGuildMemberAdd(uint32 guild_id, uint32 level, uint32 _class, uint32 rank, uint32 spirit, uint32 zone_id, std::string player_name);
+	void SendGuildMemberRename(uint32 guild_id, std::string player_name, std::string new_player_name);
```
![image](https://github.com/EQEmu/Server/assets/3319450/c389557a-835f-46cc-9ea2-4cb334bd13ca)

All of these in gray are un-used - what's the story here?

neckkola replied:
These functions were added as I worked through the individual opcodes within RoF2.  The foundation for the guild code is to simply resend the entire guild list for every action.  This allows for individual packets to be sent, reducing network load considerably.  I therefore added these for future work to redo the underlying foundation to not send the entire guild list.  These are currently not yet implemented, though perhaps I should?  Or do as a phase 2?  Please let me know your thoughts.

Akkadius replied:
As long as you have intention to follow up on using them - I'm fine with them going in

neckkola replied:
Can't get rid of me that easily.  :-)

Akkadius commented on zone/tribute.cpp:
```diff
@@ -435,9 +386,239 @@ bool ZoneDatabase::LoadTributes() {
+		if (item_id1) {
+			LogInfo("Guild Tribute Item 1 is {}", item_id1);
+			const EQ::ItemInstance* inst = database.CreateItem(item_id1, 1);
+			if (inst == nullptr) {
```
`if (!inst)`

neckkola replied:
updated

Akkadius commented on zone/tribute.cpp:
```diff
@@ -435,9 +386,239 @@ bool ZoneDatabase::LoadTributes() {
+		if (item_id2) {
+			LogInfo("Guild Tribute Item 2 is {}", item_id2);
+			const EQ::ItemInstance* inst = database.CreateItem(item_id2, 1);
+			if (inst == nullptr) {
```
`if (!inst)`

neckkola replied:
Updated

Akkadius commented on zone/guild.cpp:
```diff
@@ -317,13 +362,14 @@ void EntityList::SendGuildMembers(uint32 guild_id) {
 		Client *client = it->second;
 		if (client->GuildID() == guild_id) {
-			client->SendGuildMembers();
+			//client->SendGuildMembers();
```
Commented? Is there supposed to be a TODO here ?

neckkola replied:
This was part of the work to not have to send an entire guild list for every action (member level update, public note, alt, banker, etc.)  I removed this.  I have to check this more closely as I am unsure if I re-added all the TI/UF versions.  This should stay open while I check.

Akkadius replied:
Sounds good

Akkadius commented on zone/main.cpp:
```diff
@@ -398,7 +398,7 @@ int main(int argc, char **argv)
 	}
 
-	guild_mgr.LoadGuilds();
+	//guild_mgr.LoadGuilds();
```
Why is this commented here ?

Akkadius replied:
I see you have it later in zone.cpp now. We can remove it here if not needed

neckkola replied:
Aye.  I moved it to remove the needless sending of guild data from world to zones, when they are not booted and loaded.  Now, when a zone actually loads, it will pull the most recent data from the db, and then it will start to receive updates from world.
Updated.

Akkadius commented on zone/tribute.cpp:
```diff
@@ -435,9 +386,239 @@ bool ZoneDatabase::LoadTributes() {
+			}
+			auto inst_level = d1.tiers[guild->tribute.id_1_tier].level;
+			if (m_inv[EQ::invslot::GUILD_TRIBUTE_BEGIN]) {
+				LogInfo("Guild Tribute DELETE Item in Slot 450");
```
Probably a little debuggy for `LogInfo` perhaps rescope to `LogGuilds` ?

neckkola replied:
Changed to LogGuilds

Akkadius commented on zone/tribute.cpp:
```diff
@@ -435,9 +386,239 @@ bool ZoneDatabase::LoadTributes() {
+		}
+
+		if (item_id2) {
+			LogInfo("Guild Tribute Item 2 is {}", item_id2);
```
Probably a little debuggy for `LogInfo` perhaps rescope to `LogGuilds` ?

neckkola replied:
Changed to LogGuilds

Akkadius commented on zone/tribute.cpp:
```diff
@@ -435,9 +386,239 @@ bool ZoneDatabase::LoadTributes() {
+			auto inst_level = d2.tiers[guild->tribute.id_2_tier].level;
+			if (m_inv[EQ::invslot::GUILD_TRIBUTE_BEGIN + 1]) {
+				DeleteItemInInventory(EQ::invslot::GUILD_TRIBUTE_BEGIN + 1);
+				LogInfo("Guild Tribute DELETE Item in Slot 451");
```
Probably a little debuggy for `LogInfo` perhaps rescope to `LogGuilds` ?

neckkola replied:
Changed to LogGuilds

Akkadius commented on zone/tribute.cpp:
```diff
@@ -435,9 +386,239 @@ bool ZoneDatabase::LoadTributes() {
+		uint32 item_id2 = d2.tiers[guild->tribute.id_2_tier].tribute_item_id;
+
+		if (item_id1) {
+			LogInfo("Guild Tribute Item 1 is {}", item_id1);
```
Probably a little debuggy for `LogInfo` perhaps rescope to `LogGuilds` ?

neckkola replied:
Changed to LogGuilds and also changed the LogErrors to LogGuilds

Akkadius commented on common/database/database_update_manifest.cpp:
```diff
@@ -5155,8 +5155,67 @@ ALTER TABLE `npc_emotes` DROP INDEX `emoteid`;
+ENGINE=InnoDB
+AUTO_INCREMENT=1;
+
+UPDATE guild_ranks SET title = 'Leader' WHERE `rank` = '1';
```
I don't have anything in my `guild_ranks` table locally. How is this populated for me otherwise?

neckkola replied:
My guild_ranks from a peq install has ranks for 0 thru 8 though only 0, 1, 2 have names of Member, Officer, Leader.  This structure should be created from existing guild code.  This sql updates the rankes to be RoF2 standards.  Not sure why your local would be empty, unless there are no guilds?  I may need to update for that consideration?  Please let me know.

Akkadius commented:
Hi Neckolla!

This is a TON of work and appreciate you iterating on feedback. I think we can get this wrapped up very soon if we keep at it. 

I wanted to leave some notes for you in the morning. We've already discussed a ton of things already and addressed them so I won't rehash those.

**Items**

The items you have listed in your dump with custom ID ranges, we can pull from 13th floor using the intended Live ID. @joligario has recently done work bringing our data scraper up to speed [recently](https://github.com/EQEmu/Server/issues/3701) - we have some decisions to make there - but for items that we are straight up missing in PEQ - it's not a big deal to get the ones for tributes pulled in - we just need to coordinate what those are and we can make it happen.

**Formatting**

There were lots of wildly *off* spacing (tabs vs space translation) things that can be caused by an editor practically everywhere. I did most of the legwork just pushing up all the formatting things I found instead of leaving comments on them all to save you headache and time. You can see the commits if you're curious what. Also - I see you are manually aligning variable assignments with what seems like yourself manually. Clion IDE is doing the alignment for us and once you have code settings, it will automatically handle code formatting beautifully. We have open source licenses to allocate to regular developers to the project, so if you are interested I can hook you up with Clion and share the formatting settings.

**General**

* I've rebased the branch as we've had conflicts with master for you - to also help with iteration speed on your end to keep you lighter and focused on other parts of review and testing
* We are validating `m_db` in every database call - we just need to validate once when we construct the guild manager. If there is a database problem, it will be made very clear at either the `dbcore` level and if we're missing the database pointer, the app will crash on bootup. We don't need to validate for every call - it adds a lot of extra unnecessary logging as well. I do appreciate the practice of adding validation though.
* There's lots of large amounts unused code functions, which is surprising. I'm curious your thoughts there and I've left my comments inline in the code in client / entity headers
* We already spoke offline about the `LoadGuilds` N*3 problem which is now resolved. One thing we should make sure is that the counts for permissions line up as you expect so its not repeatedly re-inserting records over and over.
* We don't use QS logging for player events anymore, we have a whole new pipeline for handling player events https://docs.eqemu.io/server/logging-system/player-event-logging/

Everything else currently is in open comments.

It's very late, daughter has been up miserable and sick so I'm heading to bed for now. 

I'll do more passes tomorrow. 🫡

Kinglykrab commented on common/patches/uf.cpp:
```diff
@@ -1189,14 +1189,14 @@ namespace UF
 		unsigned char *__emu_buffer = in->pBuffer;
 		char *InBuffer = (char *)__emu_buffer;
-		uint32 HighestGuildID = 0;
+		uint32 actual_no_guilds = 0;
```
`guild_count`? Naming seems a bit obtuse.

neckkola replied:
Just above there is a variable called NumberOfGuilds, so I decided to use ActualNoOfGuilds, then converted to the standard so landed on actual_no_guilds.

Kinglykrab replied:
Gotcha. `actual_no_guilds` just seemed weird at first.

neckkola commented:
Good morning.  Updated the last of the comments (I think I hit them all), removed the redundant m_db checks, and updated logging to the new pipeline.

A few outstanding items
- Item IDs
- guild_ranks db questions
- confirmation for opcodes/phase 2 which is the unused code items.
- ?? anything I missed?

neckkola commented:
Good morning.  I have updated to use the single opcodes for the guild functions for Ti/UF and RoF2.  So phase 2 should be done.  I have tested and things look good.  Also, several code updates for readability.  Have a great day.

Kinglykrab commented:
Did quite a bit of testing here, seems to be good to go other than some weirdness with `Notes`/`Tribute` tab not properly updating online/offline status immediately when someone logs in/out. If we can find a way to fix this (maybe send a full guild update each time) then I think we'll be ready to merge this in.

hgtw commented on zone/client_packet.cpp:
```diff
@@ -852,9 +844,48 @@ void Client::CompleteConnect()
+			guild_mgr.GetCharInfo(CharacterID(), gci);
+
+			data->char_id        = CharacterID();
+			data->command        = 0x2fb;
```
What is 0x2fb

hgtw commented on zone/worldserver.cpp:
```diff
@@ -3382,6 +3393,225 @@ void WorldServer::HandleMessage(uint16 opcode, const EQ::Net::Packet &p)
+		GuildTributeOptInOutReply_Struct* data = (GuildTributeOptInOutReply_Struct*)outapp->pBuffer;
+
+		data->guild_id = in->guild_id;
+		strncpy(data->player_name, in->player_name, strlen(in->player_name));
```
`strn0cpy(data->player_name, in->player_name, sizeof(data->player_name));`

hgtw commented on zone/client_packet.cpp:
```diff
@@ -7712,6 +7739,14 @@ void Client::Handle_OP_GuildBank(const EQApplicationPacket *app)
 
+		if (!guild_mgr.CheckPermission(GuildID(), GuildRank(), GUILD_ACTION_BANK_WITHDRAW_ITEMS))
+		{
+			Message(Chat::Red, "You do not have permission to withdraw.");
```
Is there an eqstr for this? Or is this a custom fallback?

fryguy503 replied:
6107 You do not have permission to withdraw that item.

hgtw commented on zone/client_packet.cpp:
```diff
@@ -16542,3 +16697,300 @@ void Client::ReloadExpansionProfileSetting()
+		}
+	}
+	safe_delete(outapp)
+	return;
```
return at the end of `void` function isn't necessary, think I've seen this in a couple other functions

hgtw commented on zone/client_packet.cpp:
```diff
@@ -16542,3 +16697,300 @@ void Client::ReloadExpansionProfileSetting()
+		GuildTributeUpdate *out = (GuildTributeUpdate *) outapp->pBuffer;
+		out->guild_id = GuildID();
+		out->favor    = guild->tribute.favor;
+		strncpy(out->player_name, GetCleanName(), strlen(GetCleanName()));
```
`strn0cpy(out->player_name, GetCleanName(), sizeof(out->player_name)));`

neckkola replied:
Updated.

hgtw commented on zone/client_packet.cpp:
```diff
@@ -16542,3 +16697,300 @@ void Client::ReloadExpansionProfileSetting()
+		GuildTributeUpdate *out = (GuildTributeUpdate *) outapp->pBuffer;
+		out->guild_id = GuildID();
+		out->favor    = guild->tribute.favor;
+		strncpy(out->player_name, GetCleanName(), strlen(GetCleanName()));
```
`strn0cpy(out->player_name, GetCleanName(), sizeof(out->player_name)));`

neckkola replied:
Updated.

hgtw commented on zone/guild.cpp:
```diff
@@ -121,28 +121,49 @@ void Client::SendGuildRanks()
+			GuildUpdateUCPStruct* gucp = (GuildUpdateUCPStruct*)outapp->pBuffer;
+
+			gucp->payload.rank_name.rank = i;
+			strcpy(gucp->payload.rank_name.rank_name, guild->rank_names[i].c_str());
```
`strn0cpy(gucp->payload.rank_name.rank_name, guild->rank_names[i].c_str(), sizeof(gucp->payload.rank_name.rank_name));`

neckkola replied:
Updated.

hgtw commented on zone/client_packet.cpp:
```diff
@@ -7953,124 +8026,169 @@ void Client::Handle_OP_GuildInvite(const EQApplicationPacket *app)
+							sizeof(GuildPromoteStruct));
+						GuildPromoteStruct *gps   = (GuildPromoteStruct *) outapp->pBuffer;
+						strcpy(gps->name, gc->myname);
+						strcpy(gps->target, gc->othername);
```
`strn0cpy(gps->name, gc->myname, sizeof(gps->name));`
`strn0cpy(gps->target, gc->othername, sizeof(gps->target));`

neckkola replied:
Updated.

hgtw commented on zone/guild.cpp:
```diff
@@ -317,36 +362,93 @@ void EntityList::SendGuildMembers(uint32 guild_id) {
+	outgj->level    = gj->level;
+	outgj->rank     = gj->rank;
+	outgj->zoneid   = gj->zoneid;
 	strcpy(outgj->name, gj->name);
```
I know this is unchanged but let's add a little protection since it's fixed size
`strn0cpy(outgj->name, gj->name, sizeof(outgj->name));`

hgtw reviewed (commented):
I wasn't able to finish I'll look at it more tomorrow

hgtw commented on zone/worldserver.cpp:
```diff
@@ -3382,6 +3393,225 @@ void WorldServer::HandleMessage(uint16 opcode, const EQ::Net::Packet &p)
+		GuildTributeOptInOutReply_Struct* data = (GuildTributeOptInOutReply_Struct*)outapp->pBuffer;
+
+		data->guild_id = in->guild_id;
+		strncpy(data->player_name, in->player_name, strlen(in->player_name));
```
`strn0cpy(data->player_name, in->player_name, sizeof(data->player_name));`

joligario commented:
[GuildTributes_2023_09_13_Complete_Updated.zip](https://github.com/EQEmu/Server/files/13933447/GuildTributes_2023_09_13_Complete_Updated.zip)

Updated item ids to match main database. No need to add the new items as we have a pull for it from 13th Floor already.

neckkola commented on common/guild_base.cpp:
```diff
@@ -27,882 +35,864 @@
+	info->motd          = std::move(guild_motd);
+	info->motd_setter   = std::move(motd_setter);
+	info->url           = std::move(URL);
+	info->channel       = std::move(Channel);
```
Updated

neckkola commented on common/database/database_update_manifest.cpp:
```diff
@@ -5162,8 +5162,67 @@ ALTER TABLE `tasks`
+	`guild_id` INT(11) NOT NULL DEFAULT '0',
+	`permission` INT(11) NOT NULL DEFAULT '0',
+	PRIMARY KEY (`id`) USING BTREE
+)
```
Updated

neckkola commented on common/guild_base.cpp:
```diff
@@ -27,882 +35,864 @@
+	}
 
-	return(true);
+	auto guild_id = GetGuildIDByCharacterID(charid);
```
Good catch, removed.

neckkola commented on common/patches/rof2.cpp:
```diff
@@ -2414,16 +2433,16 @@ namespace RoF2
+		outapp->WriteUInt8(1);				// Unknown
 		outapp->WriteUInt8(emu->pvp);
-		outapp->WriteUInt8(0);				// Unknown
+		outapp->WriteUInt8(2);				// Unknown
```
Yes, was testing and not needed for final result.  Reverted to original values.

neckkola commented on common/patches/rof2.cpp:
```diff
@@ -2464,8 +2483,10 @@ namespace RoF2
+//			outapp->WriteUInt32(0xffffffff);
+//			outapp->WriteUInt32(0);
+			outapp->WriteUInt32(60);
+			outapp->WriteUInt32(1);
```
Yes, was testing and not needed for final result.  Reverted to original values.

neckkola commented on zone/guild.cpp:
```diff
@@ -121,28 +121,49 @@ void Client::SendGuildRanks()
-				guuacs->Unknown0089[2] = 0x00;
-				FastQueuePacket(&outapp);
-				i++;
+					auto outapp = new EQApplicationPacket(OP_GuildUpdateURLAndChannel,
```
There are several references to OP_GuildUpdateURLandChannel in areas outside of these updates.  I am happy to make the change though want to ensure that the intended objective and resulting outcome is as expected.  For now, I have not made any changes in this area.

neckkola commented on common/eq_packet_structs.h:
```diff
@@ -1754,6 +1792,14 @@ struct PopupResponse_Struct {
+    GuildUpdateChannel      = 1,
+    GuildUpdateRanks        = 4,
+    GuildUpdatePermissions  = 5
+};
```
Sounds good.  No changes.

neckkola commented on zone/guild.cpp:
```diff
@@ -317,36 +362,93 @@ void EntityList::SendGuildMembers(uint32 guild_id) {
+		auto                  client_list = entity_list.GetClientList();
+		for (auto const       &c: client_list) {
+			CharGuildInfo cgi;
+			guild_mgr.GetCharInfo(c.second->CharacterID(), cgi);
```
Removed.

neckkola commented on zone/guild.cpp:
```diff
@@ -317,36 +362,93 @@ void EntityList::SendGuildMembers(uint32 guild_id) {
+
+	for (auto const &c: g_members()) {
+		CharGuildInfo cgi;
+		guild_mgr.GetCharInfo(c->CharacterID(), cgi);
```
The guild info for characters is not loaded in memory and each time it is required, it is retrieved form the db.  I considered at the start of the work to move it into memory though was trying to follow the principle of not modifying the existing guild foundational pieces.  We may be to the point now to place this inside of an entity_list and manipulate in that pattern.

neckkola commented on zone/guild.cpp:
```diff
@@ -409,3 +511,458 @@ void Client::GuildChangeRank(const char* name, uint32 guild_id, uint32 oldrank,
+
+	auto outapp = new EQApplicationPacket(OP_GuildMemberList, len);
+	outapp->pBuffer = data;
+	data = nullptr;
```
Updated.

neckkola commented on zone/guild.cpp:
```diff
@@ -409,3 +511,458 @@ void Client::GuildChangeRank(const char* name, uint32 guild_id, uint32 oldrank,
+	struct GuildRenameGuild_Struct {
+		/*000*/ uint32 guild_id;
+		/*004*/ char   new_guild_name[64];
+	};
```
Moved all to eq_packets

neckkola commented on zone/guild_mgr.cpp:
```diff
@@ -396,230 +283,364 @@ bool ZoneGuildManager::VerifyAndClearInvite(uint32 char_id, uint32 guild_id, uin
+					auto outapp = new EQApplicationPacket(OP_GuildUpdateURLAndChannel,sizeof(GuildUpdateUCPStruct));
+					GuildUpdateUCPStruct *gucp  = (GuildUpdateUCPStruct *) outapp->pBuffer;
+					gucp->payload.rank_name.rank = s->rank;
+					strcpy(gucp->payload.rank_name.rank_name, s->rank_name);
```
Updated.

neckkola commented on zone/guild_mgr.cpp:
```diff
@@ -1454,3 +1475,345 @@ bool GuildBankManager::AllowedToWithdraw(uint32 GuildID, uint16 Area, uint16 Slo
+		{
+			gucp->payload.rank_name.rank = i;
+			strcpy(gucp->payload.rank_name.rank_name, guild->second->rank_names[i].c_str());
+			gucp->action = 4;
```
Updated.

neckkola commented on zone/guild_mgr.cpp:
```diff
@@ -1454,3 +1475,345 @@ bool GuildBankManager::AllowedToWithdraw(uint32 GuildID, uint16 Area, uint16 Slo
+			return true;
+		}
+	}
+	return false;
```
Updated.

neckkola commented on zone/guild.cpp:
```diff
@@ -409,3 +511,458 @@ void Client::GuildChangeRank(const char* name, uint32 guild_id, uint32 oldrank,
+		/*032*/ uint32 zone_id;
+		/*036*/ uint32 last_on;
+		/*040*/ char   player_name[64];
+	};
```
Moved

neckkola commented on zone/guild.cpp:
```diff
@@ -409,3 +511,458 @@ void Client::GuildChangeRank(const char* name, uint32 guild_id, uint32 oldrank,
+		/*000*/ uint32 guild_id;
+		/*004*/ char   player_name[64];
+		/*068*/ char   new_player_name[64];
+	};
```
moved

neckkola commented on zone/guild.cpp:
```diff
@@ -409,3 +511,458 @@ void Client::GuildChangeRank(const char* name, uint32 guild_id, uint32 oldrank,
+	struct GuildMemberDelete_Struct {
+		/*000*/ uint32 guild_id;
+		/*004*/ char   player_name[64];
+	};
```
moved

neckkola commented on zone/guild.cpp:
```diff
@@ -409,3 +511,458 @@ void Client::GuildChangeRank(const char* name, uint32 guild_id, uint32 oldrank,
+		/*000*/ uint32 guild_id;
+		/*004*/ char   player_name[64];
+		/*068*/ uint32 level;
+	};
```
moved

neckkola commented on zone/guild.cpp:
```diff
@@ -409,3 +511,458 @@ void Client::GuildChangeRank(const char* name, uint32 guild_id, uint32 oldrank,
+		/*004*/ uint32 rank;
+		/*008*/ char   player_name[64];
+		/*072*/ uint32 alt_banker; //Banker/Alt bit 00 - none 10 - Alt 11 - Alt and Banker 01 - Banker.  Banker not functional for RoF2+
+	};
```
Correct, translators look after it

neckkola commented on zone/guild.cpp:
```diff
@@ -409,3 +511,458 @@ void Client::GuildChangeRank(const char* name, uint32 guild_id, uint32 oldrank,
+		/*000*/ uint32 guild_id;
+		/*004*/ char   player_name[64];
+		/*068*/ char   public_note[256]; //RoF2 256
+	};
```
moved

neckkola commented on zone/guild.cpp:
```diff
@@ -409,3 +511,458 @@ void Client::GuildChangeRank(const char* name, uint32 guild_id, uint32 oldrank,
+		/*068*/ uint32 zone_id;
+		/*072*/ uint32 last_on;
+		/*076*/ uint32 offline_mode; //1 Offline
+	};
```
moved

neckkola commented on zone/guild.cpp:
```diff
@@ -409,3 +511,458 @@ void Client::GuildChangeRank(const char* name, uint32 guild_id, uint32 oldrank,
+	struct GuildRenameGuild_Struct {
+		/*000*/ uint32 guild_id;
+		/*004*/ char   new_guild_name[64];
+	};
```
moved

neckkola commented on zone/guild.cpp:
```diff
@@ -409,3 +511,458 @@ void Client::GuildChangeRank(const char* name, uint32 guild_id, uint32 oldrank,
+void EntityList::SendGuildMemberAdd(
+	uint32 guild_id,
+	uint32 level,
+	uint32 _class,
```
Updated

neckkola replied:
updated

neckkola commented on zone/guild.cpp:
```diff
@@ -409,3 +511,458 @@ void Client::GuildChangeRank(const char* name, uint32 guild_id, uint32 oldrank,
+void Client::SendGuildMemberAdd(
+	uint32 guild_id,
+	uint32 level,
+	uint32 _class,
```
updated

neckkola commented on zone/guild.cpp:
```diff
@@ -409,3 +511,458 @@ void Client::GuildChangeRank(const char* name, uint32 guild_id, uint32 oldrank,
+
+			c.second->QueuePacket(outapp);
+			safe_delete(outapp);
+		}
```
Modified a bit more as the client version also did the dirtylist check.

neckkola commented on zone/guild.cpp:
```diff
@@ -409,3 +511,458 @@ void Client::GuildChangeRank(const char* name, uint32 guild_id, uint32 oldrank,
+	out->level    = level;
+	out->zone_id  = zone_id;
+	out->rank     = rank;
+	out->spirt    = 0;
```
I have set to 0 as I have not determined its actual impact.  Any insight you have would be appreciated

hgtw replied:
See comment on this in eq_packet_structs `GuildMemberAdd_Struct`

neckkola replied:
Updated to use spirit

neckkola commented on zone/entity.cpp:
```diff
@@ -5850,3 +5851,49 @@ void EntityList::DamageArea(
+			}
+			else {
+				c.second->SendAppearancePacket(AT_GuildShow, 1, false, true, nullptr, true);
+			}
```
I will take a look at this closer.  Stay tuned.

neckkola replied:
Updated the overall process for your review.  It now sends to all as noted that live does.

neckkola commented on zone/guild.cpp:
```diff
@@ -409,3 +511,458 @@ void Client::GuildChangeRank(const char* name, uint32 guild_id, uint32 oldrank,
+		/*000*/ uint32 guild_id;
+		/*004*/ char   player_name[64];
+		/*068*/ char   new_player_name[64];
+	};
```
Updated

neckkola commented on zone/guild.cpp:
```diff
@@ -409,3 +511,458 @@ void Client::GuildChangeRank(const char* name, uint32 guild_id, uint32 oldrank,
+		/*068*/ uint32 zone_id;
+		/*072*/ uint32 last_on;
+		/*076*/ uint32 offline_mode; //1 Offline
+	};
```
updated

Kinglykrab commented on zone/guild.cpp:
```diff
@@ -409,3 +510,349 @@ void Client::GuildChangeRank(const char* name, uint32 guild_id, uint32 oldrank,
+	strn0cpy(out->player_name, player_name.c_str(), sizeof(out->player_name));
+
+	QueuePacket(outapp);
+	safe_delete(outapp)
```
Missing `;`.

neckkola replied:
I have put them all back.  Anything else, please let me know.

Kinglykrab commented on zone/guild.cpp:
```diff
@@ -409,3 +510,349 @@ void Client::GuildChangeRank(const char* name, uint32 guild_id, uint32 oldrank,
+	strn0cpy(out->player_name, player_name.c_str(), sizeof(out->player_name));
+
+	QueuePacket(outapp);
+	safe_delete(outapp)
```
Missing `;`.

neckkola replied:
I have put them all back.  Anything else, please let me know.

Kinglykrab commented on zone/guild.cpp:
```diff
@@ -409,3 +510,349 @@ void Client::GuildChangeRank(const char* name, uint32 guild_id, uint32 oldrank,
+	strn0cpy(out->player_name, player_name.c_str(), sizeof(out->player_name));
+
+	QueuePacket(outapp);
+	safe_delete(outapp)
```
Missing `;`.

neckkola replied:
I have put them all back.  Anything else, please let me know.

Kinglykrab commented on zone/guild.cpp:
```diff
@@ -409,3 +510,349 @@ void Client::GuildChangeRank(const char* name, uint32 guild_id, uint32 oldrank,
+	strn0cpy(out->public_note, public_note.c_str(), sizeof(out->public_note));
+
+	QueuePacket(outapp);
+	safe_delete(outapp)
```
Missing `;`.

neckkola replied:
I have put them all back.  Anything else, please let me know.

Kinglykrab commented on zone/guild.cpp:
```diff
@@ -409,3 +510,349 @@ void Client::GuildChangeRank(const char* name, uint32 guild_id, uint32 oldrank,
+	strn0cpy(out->player_name, player_name.c_str(), sizeof(out->player_name));
+
+	QueuePacket(outapp);
+	safe_delete(outapp)
```
Missing `;`.

neckkola replied:
I have put them all back.  Anything else, please let me know.

Kinglykrab commented on zone/guild.cpp:
```diff
@@ -409,3 +510,349 @@ void Client::GuildChangeRank(const char* name, uint32 guild_id, uint32 oldrank,
+	strn0cpy(out->new_guild_name, new_guild_name.c_str(), sizeof(out->new_guild_name));
+
+	QueuePacket(outapp);
+	safe_delete(outapp)
```
Missing `;`.

neckkola replied:
I have put them all back.  Anything else, please let me know.

Kinglykrab commented on zone/guild.cpp:
```diff
@@ -409,3 +510,349 @@ void Client::GuildChangeRank(const char* name, uint32 guild_id, uint32 oldrank,
+			strn0cpy(out->new_player_name, new_player_name.c_str(), sizeof(out->new_player_name));
+
+			c.second->QueuePacket(outapp);
+			safe_delete(outapp)
```
Missing `;`.

neckkola replied:
I have put them all back.  Anything else, please let me know.

Kinglykrab commented on zone/guild.cpp:
```diff
@@ -409,3 +510,349 @@ void Client::GuildChangeRank(const char* name, uint32 guild_id, uint32 oldrank,
+			strn0cpy(out->player_name, player_name.c_str(), sizeof(out->player_name));
+
+			c.second->QueuePacket(outapp);
+			safe_delete(outapp)
```
Missing `;`.

neckkola replied:
I have put them all back.  Anything else, please let me know.

Kinglykrab commented on zone/guild.cpp:
```diff
@@ -409,3 +510,349 @@ void Client::GuildChangeRank(const char* name, uint32 guild_id, uint32 oldrank,
+			strn0cpy(out->player_name, player_name.c_str(), sizeof(out->player_name));
+
+			c.second->QueuePacket(outapp);
+			safe_delete(outapp)
```
Missing `;`.

neckkola replied:
I have put them all back.  Anything else, please let me know.

Kinglykrab commented on zone/guild.cpp:
```diff
@@ -409,3 +510,349 @@ void Client::GuildChangeRank(const char* name, uint32 guild_id, uint32 oldrank,
+			strn0cpy(out->player_name, player_name.c_str(), sizeof(out->player_name));
+
+			c.second->QueuePacket(outapp);
+			safe_delete(outapp)
```
Missing `;`.

neckkola replied:
I have put them all back.  Anything else, please let me know.

Kinglykrab commented on zone/guild.cpp:
```diff
@@ -409,3 +510,349 @@ void Client::GuildChangeRank(const char* name, uint32 guild_id, uint32 oldrank,
+			strn0cpy(out->public_note, public_note.c_str(), sizeof(out->public_note));
+
+			c.second->QueuePacket(outapp);
+			safe_delete(outapp)
```
Missing `;`.

neckkola replied:
I have put them all back.  Anything else, please let me know.

Kinglykrab commented on zone/guild.cpp:
```diff
@@ -409,3 +510,349 @@ void Client::GuildChangeRank(const char* name, uint32 guild_id, uint32 oldrank,
+			strn0cpy(out->player_name, player_name.c_str(), sizeof(out->player_name));
+
+			c.second->QueuePacket(outapp);
+			safe_delete(outapp)
```
Missing `;`.

neckkola replied:
I have put them all back.  Anything else, please let me know.

Kinglykrab commented on zone/guild.cpp:
```diff
@@ -409,3 +510,349 @@ void Client::GuildChangeRank(const char* name, uint32 guild_id, uint32 oldrank,
+			strn0cpy(out->new_guild_name, new_guild_name.c_str(), sizeof(out->new_guild_name));
+
+			c.second->QueuePacket(outapp);
+			safe_delete(outapp)
```
Missing `;`.

neckkola replied:
I have put them all back.  Anything else, please let me know.

Kinglykrab commented on zone/guild_mgr.cpp:
```diff
@@ -1454,3 +1475,338 @@ bool GuildBankManager::AllowedToWithdraw(uint32 GuildID, uint16 Area, uint16 Slo
+	strn0cpy(sr->member_name, player_name, sizeof(sr->member_name));
+
+	worldserver.SendPacket(outapp);
+	safe_delete(outapp)
```
Missing `;`.

neckkola replied:
I removed it as it is within the macro and CLion highlights it to be removed.

neckkola replied:
I have put them all back.  Anything else, please let me know.

Kinglykrab commented on zone/guild_mgr.cpp:
```diff
@@ -1454,3 +1475,338 @@ bool GuildBankManager::AllowedToWithdraw(uint32 GuildID, uint16 Area, uint16 Slo
+	sr->guild_id  = guild_id;
+
+	worldserver.SendPacket(outapp);
+	safe_delete(outapp)
```
Missing `;`.

neckkola replied:
I have put them all back.  Anything else, please let me know.

Kinglykrab commented on zone/mob.cpp:
```diff
@@ -3982,33 +3982,26 @@ uint8 Mob::GetDefaultGender(uint16 in_race, uint8 in_gender) {
+		if (IsClient() && CastToClient()->GuildID()) {
+			entity_list.QueueClientsNotInGuild(this, outapp, iIgnoreSelf, CastToClient()->GuildID());
+		}
+	}else if (WholeZone) {
```
`} else if`

neckkola replied:
This function is under review with HG so will update once finalized there.

Kinglykrab commented on zone/mob.cpp:
```diff
@@ -3982,33 +3982,26 @@ uint8 Mob::GetDefaultGender(uint16 in_race, uint8 in_gender) {
+		}
+	}else if (WholeZone) {
+		entity_list.QueueClients(this, outapp, iIgnoreSelf);
+	}else if (specific_target != nullptr) {
```
`} else if (specific_target) {`

Kinglykrab commented on zone/mob.cpp:
```diff
@@ -3982,33 +3982,26 @@ uint8 Mob::GetDefaultGender(uint16 in_race, uint8 in_gender) {
 		CastToClient()->QueuePacket(outapp, false, Client::CLIENT_CONNECTED);
 	}
-
+	
```
Extra tab?

Kinglykrab commented on zone/mob.h:
```diff
@@ -1025,8 +1025,8 @@ class Mob : public Entity {
 	static uint32 GetAppearanceValue(EmuAppearance in_appearance);
-	void SendAppearancePacket(uint32 type, uint32 value, bool whole_zone = true, bool ignore_self = false, Client* target = nullptr);
-	void SetAppearance(EmuAppearance app, bool ignore_self = true);
+    void SendAppearancePacket(uint32 type, uint32 value, bool WholeZone = true, bool iIgnoreSelf = false, Client* specific_target = nullptr, bool WholeZoneMinusMyGuild = false);
```
Why were these reverted to non `snake_case`?

neckkola replied:
The function is under review elsewhere and will update once finalized.

neckkola replied:
New updates here for review

Kinglykrab commented on zone/tribute.cpp:
```diff
@@ -435,9 +386,243 @@ bool ZoneDatabase::LoadTributes() {
+	}
+
+	QueuePacket(outapp);
+	safe_delete(outapp)
```
Missing `;`.

neckkola replied:
I have put them all back.  Anything else, please let me know.

Kinglykrab commented on zone/tribute.cpp:
```diff
@@ -435,9 +386,243 @@ bool ZoneDatabase::LoadTributes() {
+	out->quantity = in->quantity;
+
+	QueuePacket(outapp);
+	safe_delete(outapp)
```
Missing `;`.

neckkola replied:
I have put them all back.  Anything else, please let me know.

Kinglykrab commented on zone/tribute.cpp:
```diff
@@ -435,9 +386,243 @@ bool ZoneDatabase::LoadTributes() {
+
+	QueuePacket(outapp);
+	safe_delete(outapp)
+
```
Extra newline.

neckkola replied:
I have put them all back.  Anything else, please let me know.

Kinglykrab commented on zone/worldserver.cpp:
```diff
@@ -3382,6 +3393,225 @@ void WorldServer::HandleMessage(uint16 opcode, const EQ::Net::Packet &p)
+		out->tribute_id_2_tier = guild->tribute.id_2_tier;
+
+		entity_list.QueueClientsGuild(outapp, in->guild_id);
+		safe_delete(outapp)
```
Missing `;`.

neckkola replied:
I have put them all back.  Anything else, please let me know.

hgtw commented on common/eq_packet_structs.h:
```diff
@@ -1671,6 +1671,68 @@ struct GuildUpdate_Struct {
+	/*016*/ uint32 level;
+	/*020*/ uint32 _class;
+	/*024*/ uint32 rank;
+	/*028*/ uint32 spirt; //not confirmed single byte 0x000000FF
```
This field is a padded bool that shadows the `Display Guild Name` permission of the guild for the added player's rank (Recruit?) to show the guild name sprite above the player's head.

It might be unused now in favor of the AT_GuildShow stat change msg that updates the name sprite explicitly but would need to double check and to be safe if it's not complicated to add we should fill it with the guild's current permission value for recruits.

neckkola replied:
hg, I looked at this closer and GuildShow is an RoF item and that type is not available in UF.  Any suggestions on backwards compatibility please let me know.  Thanks!

hgtw replied:
There's probably not a way to implement the permission for older clients without getting hacky and sending a fake guild id -1 update for their player to everyone (which might break something else, would have to look closer). I'd ignore it for now on older clients and just send it for rof2.

This specific field doesn't appear to be used anyway from what I see and might just be for server side, but it should be easy to update it to match what live does. The actual display update for the permission change is done through the GuildShow update. The hacky solution for older clients can be investigated in another patch later.

neckkola replied:
Thanks.  I will update with this approach. Thanks again for the insight.

hgtw replied:
Looking at it a little more for GuildShow specifically, older clients could bail out in the translators when handling `OP_SpawnAppearance` if the type isn't supported (don't forget to delete the packet if doing this).  We might need to do this eventually for other (non-guild related) types.

neckkola replied:
@hgtw, refactored the sendappearance functionality for GuildShow and updated the translators for UF/TI to discard the that type.  Appreciate your thoughts if you have a moment.

neckkola replied:
Updated the db index as well

neckkola replied:
and below spirit item

neckkola commented on zone/guild_mgr.cpp:
```diff
@@ -396,230 +283,364 @@ bool ZoneGuildManager::VerifyAndClearInvite(uint32 char_id, uint32 guild_id, uin
+				);
+				safe_delete(outapp);
+
+				if (sgpus->function_id == GuildUpdateRanks) {
```
You are correct, they both happy to be 4.  Updated.

hgtw commented on common/eq_packet_structs.h:
```diff
@@ -1671,6 +1671,68 @@ struct GuildUpdate_Struct {
+	/*016*/ uint32 level;
+	/*020*/ uint32 class_;
+	/*024*/ uint32 rank_;
+	/*028*/ uint32 spirit; //not confirmed single byte 0x000000FF
```
Let's rename this field to indicate it's for the permission of showing the guild name

neckkola replied:
Updated

hgtw commented on zone/guild.cpp:
```diff
@@ -409,3 +512,349 @@ void Client::GuildChangeRank(const char* name, uint32 guild_id, uint32 oldrank,
+	uint32 level,
+	uint32 class_,
+	uint32 rank_,
+	uint32 spirit,
```
Rename this one too to match a struct field rename

neckkola replied:
Updated

hgtw reviewed (commented):
For comments on using zone spawn packets instead of the appearance updates on client connection: if it's too difficult to change these now it could be done in a followup if you don't want to tackle it but it would probably significantly reduce packets on client connections for guild updates. From only looking at it I'm not sure if it would be an easy switch or not.

neckkola commented on zone/client_packet.cpp:
```diff
@@ -899,6 +930,9 @@ void Client::CompleteConnect()
 	entity_list.ScanCloseMobs(close_mobs, this, true);
 
+	entity_list.SendToGuildTitleDisplay(this);
+	SendGuildSpawnAppearance();
```
Thanks.  I have been thinking about this and think it wise to update the zone spawn packets and spawn struct as it may resolve the guild window as well.  That said, I can spend some time taking a look.  I would like to dig deeper on the spawn functionality.  Thanks hg.

neckkola replied:
Updated to use the zone spawn for guild display and removed unused functions.  Thanks for the advice hg as this will also benefit my trader/buyer/bazaar work.

hgtw replied:
Looks like the wrong line was removed here:
`entity_list.ScanCloseMobs(close_mobs, this, true);`
should be added back
`SendGuildSpawnAppearance();`
should be removed after the zone spawn changes

neckkola commented on zone/entity.cpp:
```diff
@@ -5905,3 +5906,45 @@ void EntityList::DamageArea(
+			client.second->SendAppearancePacket(AppearanceType::GuildShow, display, false, false, c);
+		}
+	}
+}
```
Removed

neckkola commented on zone/client_packet.cpp:
```diff
@@ -857,9 +849,48 @@ void Client::CompleteConnect()
+		}
+
+		SendAppearancePacket(AppearanceType::GuildID, GuildID(), true);
+		SendAppearancePacket(AppearanceType::GuildRank, GuildRank(), false);
```
Removed

neckkola commented on zone/entity.cpp:
```diff
@@ -5905,3 +5906,45 @@ void EntityList::DamageArea(
+	if (c) {
+		if (c->ClientVersion() < EQ::versions::ClientVersion::RoF) {
+			return;
+		}
```
It did seem to be ok.  I have now removed this function given the zone spawn aspects.

neckkola commented:
@hg Thanks for the comments regarding zone spawns.  I had a look and have utilized that functionality to remove the specific sendappearance packets.  This has allowed for several functions to be removed.  Please let me know if you see anything else.  Once ok, I will rerun my tests.

hgtw commented:
Nothing else jumps out with the zone packet stuff so I think you're good to run tests there after fixing the `ScanCloseMobs`/`SendGuildSpawnAppearance` line removal swap issue.

The only other major issues left I think are the world comments (mainly the possible `GetCLE()` nullptr pointer issue and consolidating all the duplicate booted zone server packet sends). There's also a few more `strn0cpy` changes for overflow safety that wouldn't hurt.

Outside of those the only concern would be some of the db queries that occur in loops instead of being batched for outside, but I'm not sure if those are low frequency updates or not.

neckkola commented on world/client.cpp:
```diff
@@ -2388,3 +2388,40 @@ void Client::RecordPossibleHack(const std::string& message)
+
+void Client::SendGuildTributeFavorAndTimer(uint32 favor, uint32 time_remaining) 
+{
+	auto guild = guild_mgr.GetGuildByGuildID(GetCLE()->GuildID());
```
Put a check in to be sure.

neckkola commented on world/client.cpp:
```diff
@@ -2388,3 +2388,40 @@ void Client::RecordPossibleHack(const std::string& message)
+	GuildTributeOptInOutReply_Struct* data = (GuildTributeOptInOutReply_Struct*)outapp->pBuffer;
+
+	data->guild_id = in->guild_id;
+	strncpy(data->player_name, in->player_name, 64);
```
Updated

neckkola commented on world/wguild_mgr.cpp:
```diff
@@ -133,15 +144,323 @@ void WorldGuildManager::ProcessZonePacket(ServerPacket *pack) {
+				if (r->GetZoneID() > 0) {
+					r->SendPacket(pack);
+				}
+			}
```
Created SendPacketToBootedZones and removed the duplications.

neckkola commented on world/wguild_mgr.cpp:
```diff
@@ -133,15 +144,323 @@ void WorldGuildManager::ProcessZonePacket(ServerPacket *pack) {
+			g.second->tribute.timer.Start(RuleI(Guild, TributeTime));
+
+			guild_mgr.UpdateDbGuildFavor(g.first, g.second->tribute.favor);
+			guild_mgr.UpdateDbTributeTimeRemaining(g.first, RuleI(Guild, TributeTime));
```
Great question.  Currently it will hit the db when the Guild Tribute timer expires (every 10 mins) or when the RefreshInterval (default is 3 mins) expires.  Let me know what you think.

hgtw replied:
If the interval is that slow then it should be fine here. It can be changed in a followup if it proves to be an issue on servers with lots of guilds.

neckkola commented:
Good morning @hgtw 

I have reviewed the world comments and believe I have them updated to resolve the comments and re-inserted the drop scanmobs item.  Apologize for that oversight.

Also repaired a few other items that I saw, and updated all the remaining strcpy items and left some notes for the db loop items.

@Kinglykrab I believe I have the formatting items resolved as well.  

Anything further, please let me know as I may have overlooked something.

hgtw reviewed (approved):
I don't see any major issues so as long as it tests ok after the changes it looks good. Nice work.

neckkola commented:
Awesome.  I will do some testing on my end.  Would welcome others doing the same given the size and scope of the changes.

neckkola commented:
I ran through several test cases between RoF2, UF and TI.  Found a few items along the way. 
Also loaded JJ's db updates (thank you!) and all seems good there.  I have not tested all the tributes, though @fryguy503 has been running them for some time with no reports of issue.

Kinglykrab reviewed (approved):
Looks good to me, think it's good to go. Will let @Akkadius take another once over if he wants.

neckkola commented:
Do the db changes need to be pushed to peq_content?

noudess commented:
> Do the db changes need to be pushed to peq_content?

Does this mean people need db changes if we dont use currect pew content?

Akkadius commented on zone/guild_mgr.cpp:
```diff
@@ -1454,3 +1474,338 @@ bool GuildBankManager::AllowedToWithdraw(uint32 GuildID, uint16 Area, uint16 Slo
+	m.alt            = 0;
+	m.banker         = rank == GUILD_LEADER ? 1 : 0;
+	m.last_tribute   = 0;
+    m.total_tribute  = 0;
+    m.tribute_enable = 0;
+    m.rank_          = rank;
+    m.char_id        = char_id;
+    m.guild_id       = guild_id;
+    m.online         = 1;
+    m.public_note.clear();
```
Not sure what happened here

Akkadius commented on zone/client_packet.cpp:
```diff
@@ -16542,3 +16697,300 @@ void Client::ReloadExpansionProfileSetting()
+	out->char_id        = gci.char_id;
+	out->tribute_toggle = in->tribute_toggle;
+	out->command        = in->command;
+	strncpy(out->player_name, in->player, strlen(in->player));
```
Did this for Neck

Akkadius commented:
> [GuildTributes_2023_09_13_Complete_Updated.zip](https://github.com/EQEmu/Server/files/13933447/GuildTributes_2023_09_13_Complete_Updated.zip)
> 
> Updated item ids to match main database. No need to add the new items as we have a pull for it from 13th Floor already.

Added these on PEQ

Akkadius commented:
> Do the db changes need to be pushed to peq_content?

This is done
